packages/css/src/plugins/atomicify-rules.tsx

```diff
@@ -0,0 +1,215 @@
+import { plugin, Node, Declaration, decl, Container, rule, Rule, AtRule } from 'postcss';
+import { hash } from '@compiled/utils';
+
+interface PluginOpts {
+  callback?: (className: string) => void;
+}
+
+interface AtomicifyOpts extends PluginOpts {
+  selectors?: string[];
+  atRule?: string;
+  parentNode?: Container;
+}
+
+/**
+ * Returns an atomic rule class name using this form:
+ *
+ * ```
+ * "_{atrules}{selectors}{propertyname}"
+ * ```
+ *
+ * Atomic rules are always prepended with an underscore.
+ *
+ * @param group
+ * @param value
+ */
+const atomicClassName = (propName: string, value: string, opts: AtomicifyOpts) => {
+  const selectors = opts.selectors ? opts.selectors.join('') : '';
+  const group = hash(`${opts.atRule}${selectors}${propName}`).slice(0, 4);
+  const valueHash = hash(value).slice(0, 4);
+
+  return `_${group}${valueHash}`;
+};
+
+/**
+ * Returns a normalized selector.
+ * This will handle immediate dangling pseudos `:` as well as the nesting operator `&`.
+ *
+ * @param selector
+ */
+const normalizeSelector = (selector: string | undefined) => {
+  if (!selector) {
+    // Nothing to see here - return early with an empty string!
+    return '';
+  }
+
+  // We want to build a consistent selector that we will use to generate the group hash.
+  // Because of that we trim whitespace as well as removing any top level nesting "&".
+  const trimmed = selector.trim();
+
+  switch (trimmed.charAt(0)) {
+    case ':':
+      // Dangling pseudo - return immedately!
+      return trimmed;
+
+    case '&':
+      // Dangling nesting - replace it and return!
+      return trimmed.replace('&', '');
+
+    default:
+      // Must be a nested selector - add a space before it!
+      return ` ${trimmed}`;
+  }
+};
+
+/**
+ * Replaces all instances of a nesting operator `&` with the parent class name.
+ *
+ * @param selector
+ * @param parentClassName
+ */
+const replaceNestingSelector = (selector: string, parentClassName: string) => {
+  return selector.replace(/ &/g, ` .${parentClassName}`);
+};
+
+/**
+ * Builds an atomic rule selector.
+ *
+ * @param node
+ */
+const buildAtomicSelector = (node: Declaration, opts: AtomicifyOpts) => {
+  const selectors: string[] = [];
+
+  (opts.selectors || ['']).forEach((selector) => {
+    const normalizedSelector = normalizeSelector(selector);
+    const className = atomicClassName(node.prop, node.value, {
+      ...opts,
+      selectors: [normalizedSelector],
+    });
+    const replacedSelector = replaceNestingSelector(normalizedSelector, className);
+
+    selectors.push(`.${className}${replacedSelector}`);
+
+    if (opts.callback) {
+      opts.callback(className);
+    }
```

Comment:
nit: Should we call callback once `forEach` ends? I haven't checked its implementation. Just came here because of a test where we are accumulating these in an array.

Reply:
we'd have to save the class names locally as well - i figured its just easier to do this - wdyt?

Reply:
I just thought that this `callback` is getting called on every iteration and may be we can call it in the end instead. If there is no performance impact, then we are good 😃 Yups, in current implementation we don't have to save it locally. Just a nit, you can ignore 😄